### PR TITLE
[dht] concurrent queries

### PR DESCRIPTION
**Pull request by sim590**
This PR is completing some work unintentionaly left undone after the queries. This has left place to some bugs like infinite loops when one or more concurent queries would be issued before a first request would complete.

**This PR introduces the following behavior**:

When multiple queries are queued for a same hash, the search will find the closest node to which send the first queued query. This process will repeat for each queries until all queries are sent to a sufficient number of nodes.

*Note that no more than `MAX_REQUESTED_SEARCH_NODES` nodes will be solicited. However, if 100 requests are queued for a same hash, there could be up to 100x`MAX_REQUESTED_SEARCH_NODES` requests sent at the same time since we only count the solicited nodes and not the requests. This will most of the time not be true because queries are susceptible to satisfy others, hence substantially reducing the number of requests sent. @aberaud: if you think this should change, let me know before merging.*

### CODE REVIEW

**Comment by sim590** [comment]
I have not yet tested this with the latest merged commit on master. Let me rebase and check if everything works appropriately.

**Comment by sim590** [comment]
I have just checked and I have not notcied any difference in the code's behavior. 

+1

File: src/dht.cpp [hunk 1]
```diff
@@ -265,31 +265,34 @@ struct Dht::SearchNode {
...
      * @param q       The query defining the "get" operation we're referring to.
      *
      * @return true if we can send get, else false.
      */
     bool canGet(time_point now, time_point update, std::shared_ptr<Query> q = {}) const {
         /* Find request status for the given query */
-        const auto& get_status = getStatus.find(q);
+        const auto& completed_sq_status = std::find_if(getStatus.cbegin(), getStatus.cend(),
+            [&q,&update](const SyncStatus::value_type& s) {
+                return s.first and q and q->isSatisfiedBy(*s.first) and s.second and s.second->reply_time > update;
```
Comment (aberaud) [inline src/dht.cpp]: isSatifiedBy is called twice for every getStatus. Maybe there is a way to call it only once ?

Comment (aberaud) [inline src/dht.cpp]: (could be performed in a single loop ?)

File: src/dht.cpp [hunk 2]
```diff
@@ -1366,11 +1429,11 @@ unsigned Dht::Search::getNumberOfBadNodes() const {
 }
 
 time_point
-Dht::Search::getLastGetTime() const
+Dht::Search::getLastGetTime(std::shared_ptr<Query> q) const
 {
     time_point last = time_point::min();
     for (const auto& g : callbacks)
-        last = std::max(last, g.second.start);
+        last = std::max(last, (q and q->isSatisfiedBy(*g.second.query) ? g.second.start : time_point::min()));
```
Comment (aberaud) [inline src/dht.cpp]: What if q is not set

Comment (sim590) [inline src/dht.cpp]: Indeed!

File: src/dht.cpp [hunk 3]
```diff
@@ -1029,31 +1076,36 @@ Dht::searchSendGetValues(std::shared_ptr<Search> sr, SearchNode* pn, bool update
...
-        if (pn) {
-            if (not pn->canGet(now, up, query))
-                return nullptr;
-            n = pn;
-        } else {
-            for (auto& sn : sr->nodes) {
-                if (sn.canGet(now, up, query)) {
-                    n = &sn;
-                    break;
+        auto get_available_node =
```
Comment (aberaud) [inline src/dht.cpp]: flatten the lambda

File: src/dht.cpp [hunk 4]
```diff
@@ -1086,10 +1144,13 @@ Dht::searchSendGetValues(std::shared_ptr<Search> sr, SearchNode* pn, bool update
                 paginate(ws, query, n);
         }
 
-        if (not sr->isSynced(now) or cb == sr->callbacks.end())
-            break; /* only trying to find nodes, only send the oldest query */
+        /* We only try to send one request. return. */
+        return n;
```
Comment (aberaud) [inline src/dht.cpp]: could send all requests for this node ?

**Pull request merged**